Hook keras Layers and Models

yanndupis opened the pull request:
Hey @jvmancuso, 

The branch hook-keras-layers has been moved fom DL repo to OM repo. Let's review the keras hook from here instead of #14. 

Thanks

jvmancuso commented on syft_tensorflow/hook/hook.py:
```diff
@@ -98,10 +104,94 @@ def _hook_native_tensor(self, tensor_type: type, syft_type: type):
 
         # Overload auto overloaded with Torch methods
-        self._add_methods_from_native_tensor(tensor_type, syft_type)
+        exclude = [  # Overload auto overloaded with Torch methods
```
can we clean these comments up please 🙏

yanndupis replied:
This should be cleaned up by #16 if not, I will do it.

jvmancuso commented on syft_tensorflow/hook/hook.py:
```diff
@@ -147,18 +237,16 @@ def _add_registration_to___init__(
                 logic. TODO: this flag might never get used.
```
can we update this docstring to reflect changes below and remove this TODO

jvmancuso commented on syft_tensorflow/serde/serde.py:
```diff
@@ -138,6 +140,171 @@ def _detail_tf_variable(worker, tensor_tuple) -> tf.Tensor:
+    """
+
+    params = history_cb.params
+    # serde can't serialize numpy values. These values need
```
serde _can_ serialize numpy values, but right now it's in torch_serde.py -- I seem to remember that using pickle by default (which we might want to avoid), but either way I'd take another look at this and be sure this is the right approach

jvmancuso replied:
but the right approach could just be to move that ndarray simplifier/detailer into native_serde.py, which would also fix the problem

jvmancuso replied:
cause then you could just call syft.serde._simplify on the `hisotry_cb.history`

yanndupis replied:
good point, we don't need this anymore. We are now serializing numpy numbers in the main PySyft https://github.com/dropoutlabs/PySyft/pull/16. Thanks

jvmancuso commented on syft_tensorflow/serde/serde_utils.py:
```diff
@@ -0,0 +1,20 @@
+
+
+def history_numpy_to_float(history):
```
if we end up using the numpy simplifier/detailer, this should be removed

yanndupis replied:
fixed, see comment above

jvmancuso commented on syft_tensorflow/tensor/keras_layer.py:
```diff
@@ -0,0 +1,104 @@
+    checking AbstractObject.
+    """
+
+    def send(
```
let's open an issue to standardize this & `KerasModel.send` (and potentially any other objects from PySyft main we'd like to handle the same way)

jvmancuso replied:
looking further into this PR, it looked like `KerasObject` might be the right place to abstract the `send` method

yanndupis replied:
as discussed in Slack, I am not able to call send from `KerasObject` in KerasModel` and `KerasLayer`. We can revisit later. thanks

jvmancuso commented on syft_tensorflow/tensor/keras_model.py:
```diff
@@ -0,0 +1,104 @@
+import weakref
```
can we rename the folder containing this file to `syft_tensorflow/types/`, ``syft_tensorflow/classes/`, `syft_tensorflow/objects/`, or something similar but more informative? `tensor` seems to be a misnomer after this PR

yanndupis replied:
Done, thanks

yanndupis commented:
Hey @jvmancuso, I think I have addressed most of the comments, just need to get this PR merged: https://github.com/dropoutlabs/PySyft/pull/16.

Thanks
